Included path which include name of original directory are not included

### New Issue Checklist

- [X] Updated SwiftLint to the latest version
- [X] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

I have two directories next to each other. The first directory contains a Xcode project on which swiftlint is supposed to be called on from the command line. In the other directory which has the same name as the first directory, just with a small addition to the name, has other Swift files which are also supposed to be linted. Due to the overlapping name the the second directory doesn't get scanned for files to be linted. When changing the name of the second directory to something that doesn't overlap with the first name running the same command the files from the second directory get detected.

An example of this can be found here: https://github.com/Cronay/swiftlint_name_collision

Swiftlint is supposed to be called within the 'samename' directory and the 'samenamexxx' directory also should be linted according to the .swiftlint.yml. Only when I change 'samenamexxx' to (for example) 'anotherdir' and also enter this in the config file then it get's scanned for swift files and those are linted.

Background is we have a real-world project at my company where we modularize our application via cocoapods and have the umbrella-app repository named 'ios-my-company-name' and all the other pod repos which are supposed to be linted we have named 'ios-my-company-name-core' or more generally 'ios-my-company-name-<module-name>'. The reside next to each other in a directory and are included so we can import them as development pods via cocoapods. The linked repo above is just the minimal reproducable example I could find.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint

Linting Swift files in current working directory
Linting 'ViewController.swift' (1/2)
Linting 'AppDelegate.swift' (2/2)
/Users/yborgfeld/Desktop/Test/samename/LintingProject/ViewController.swift:26:1: warning: Trailing Newline Violation: Files should have a single trailing newline. (trailing_newline)
/Users/yborgfeld/Desktop/Test/samename/LintingProject/ViewController.swift:24:1: warning: Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 2. (vertical_whitespace)
/Users/yborgfeld/Desktop/Test/samename/LintingProject/AppDelegate.swift:26:1: warning: Trailing Newline Violation: Files should have a single trailing newline. (trailing_newline)
/Users/yborgfeld/Desktop/Test/samename/LintingProject/AppDelegate.swift:13:1: warning: Trailing Whitespace Violation: Lines should not have trailing whitespace. (trailing_whitespace)
/Users/yborgfeld/Desktop/Test/samename/LintingProject/AppDelegate.swift:15:1: warning: Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 4. (vertical_whitespace)
/Users/yborgfeld/Desktop/Test/samename/LintingProject/AppDelegate.swift:24:1: warning: Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 2. (vertical_whitespace)
Done linting! Found 6 violations, 0 serious in 2 files.
```

### Environment

* SwiftLint version: 0.43.1
* Installation method: Homebrew or CocoaPods (doesn't matter)
* Paste your configuration file:

```yml
included:
  - ../samename
  - ../samenamexxx
```
Xcode Version: 12.5.1

Edit: This bug also occurs when only the second directory is in the included paths.

We found a workaround by passing the included directories via the command line and remove the 'included' section from the `.swiftlint.yml` file. Seems like this is only a bug with the yaml file.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -48,6 +48,11 @@
   [SimplyDanny](https://github.com/SimplyDanny)
   [#3975](https://github.com/realm/SwiftLint/issues/3975)
 
+* Make sure that include paths prefixed with the name of the original path
+  are included in the analysis.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#3705](https://github.com/realm/SwiftLint/issues/3705)
+
 ## 0.47.1: Smarter Appliance
 
 #### Breaking

--- a/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
@@ -103,14 +103,15 @@ extension String {
 
     /// If self is a path, this method can be used to get a path expression relative to a root directory
     public func path(relativeTo rootDirectory: String) -> String {
-        var rootDirComps = rootDirectory.components(separatedBy: "/")
+        var rootDirComps = rootDirectory.bridge().standardizingPath.components(separatedBy: "/")
         let rootDirCompsCount = rootDirComps.count
+        let normalizedSelf = bridge().standardizingPath
 
         while true {
             let sharedRootDir = rootDirComps.joined(separator: "/")
-            if self == sharedRootDir || hasPrefix(sharedRootDir + "/") {
+            if normalizedSelf == sharedRootDir || normalizedSelf.hasPrefix(sharedRootDir + "/") {
                 let path = (0 ..< rootDirCompsCount - rootDirComps.count).map { _ in "/.." }.flatMap { $0 }
-                    + String(dropFirst(sharedRootDir.count))
+                    + String(normalizedSelf.dropFirst(sharedRootDir.count))
                 return String(path.dropFirst()) // Remove leading '/'
             } else {
                 rootDirComps = rootDirComps.dropLast()

--- a/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
@@ -103,9 +103,13 @@ extension String {
 
     /// If self is a path, this method can be used to get a path expression relative to a root directory
     public func path(relativeTo rootDirectory: String) -> String {
-        var rootDirComps = rootDirectory.bridge().standardizingPath.components(separatedBy: "/")
-        let rootDirCompsCount = rootDirComps.count
+        let normalizedRootDir = rootDirectory.bridge().standardizingPath
         let normalizedSelf = bridge().standardizingPath
+        if normalizedRootDir.isEmpty {
+            return normalizedSelf
+        }
+        var rootDirComps = normalizedRootDir.components(separatedBy: "/")
+        let rootDirCompsCount = rootDirComps.count
 
         while true {
             let sharedRootDir = rootDirComps.joined(separator: "/")

--- a/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
+++ b/Source/SwiftLintFramework/Extensions/String+SwiftLint.swift
@@ -108,7 +108,7 @@ extension String {
 
         while true {
             let sharedRootDir = rootDirComps.joined(separator: "/")
-            if hasPrefix(sharedRootDir) {
+            if self == sharedRootDir || hasPrefix(sharedRootDir + "/") {
                 let path = (0 ..< rootDirCompsCount - rootDirComps.count).map { _ in "/.." }.flatMap { $0 }
                     + String(dropFirst(sharedRootDir.count))
                 return String(path.dropFirst()) // Remove leading '/'

--- a/Tests/SwiftLintFrameworkTests/StringExtensionTests.swift
+++ b/Tests/SwiftLintFrameworkTests/StringExtensionTests.swift
@@ -12,5 +12,6 @@ final class StringExtensionTests: XCTestCase {
         XCTAssertEqual("../MyFolder123", "Folder/MyFolder123".path(relativeTo: "Folder/MyFolder"))
         XCTAssertEqual("../MyFolder123", "Folder/MyFolder123".path(relativeTo: "Folder/MyFolder/"))
         XCTAssertEqual("Test", "Root////Folder///Test/".path(relativeTo: "Root//Folder////"))
+        XCTAssertEqual("Root/Folder/Test", "Root/Folder/Test/".path(relativeTo: ""))
     }
 }

--- a/Tests/SwiftLintFrameworkTests/StringExtensionTests.swift
+++ b/Tests/SwiftLintFrameworkTests/StringExtensionTests.swift
@@ -9,5 +9,6 @@ final class StringExtensionTests: XCTestCase {
         XCTAssertEqual("../Test", "Root/Folder/Test".path(relativeTo: "Root/Folder/SubFolder"))
         XCTAssertEqual("../..", "Root".path(relativeTo: "Root/Folder/SubFolder"))
         XCTAssertEqual("../../OtherFolder/Test", "Root/OtherFolder/Test".path(relativeTo: "Root/Folder/SubFolder"))
+        XCTAssertEqual("../MyFolder123", "Folder/MyFolder123".path(relativeTo: "Folder/MyFolder"))
     }
 }

--- a/Tests/SwiftLintFrameworkTests/StringExtensionTests.swift
+++ b/Tests/SwiftLintFrameworkTests/StringExtensionTests.swift
@@ -10,5 +10,7 @@ final class StringExtensionTests: XCTestCase {
         XCTAssertEqual("../..", "Root".path(relativeTo: "Root/Folder/SubFolder"))
         XCTAssertEqual("../../OtherFolder/Test", "Root/OtherFolder/Test".path(relativeTo: "Root/Folder/SubFolder"))
         XCTAssertEqual("../MyFolder123", "Folder/MyFolder123".path(relativeTo: "Folder/MyFolder"))
+        XCTAssertEqual("../MyFolder123", "Folder/MyFolder123".path(relativeTo: "Folder/MyFolder/"))
+        XCTAssertEqual("Test", "Root////Folder///Test/".path(relativeTo: "Root//Folder////"))
     }
 }

